### ISSUE TITLE
Controller Priority Option (request)

### ISSUE THREAD

**Issue by Kayinnasaki** [libretro/RetroArch#6707]
A recent change to [the way controller order is handled in linux](https://github.com/libretro/RetroArch/commit/b14598657c4b813c555b6cc6c6e81ec0830621c4?diff=unified )   has turned into the bane of some builds. For example, I have a Raspberry Pi joystick build where previously the gpio joystick had the desirable behavior of being pushed back by any additional controllers plugged in. Wanna have two people play mario kart? Awesome, plug in two SNES controllers and the joystick moves to player 3 and doesn't bother anyone. Now its player 1 no matter what and the only thing I can do is log in through SSH to activate or deactivate it. =/

Obviously the current way of handling it in normal situations is much more sensible but kinda highlights the need for a feature to either specify controller priority or at the ver least, specify low priority controllers. "Hey, this is always plugged in so treat it like you would a keyboard and move it out of the way".

I feel like this would help a lot of people who do arcade or portable style builds that leverage retroatch and probably would be simply a nice feature in general.

**Comment by i30817** [comment]
The new way doesn't seem more sensible though, for the exact reason you mention.

**Comment by Kayinnasaki** [comment]
But the old behavior seemed less 'stack' and more... semi random? At least when mixing bluetooth, USB and other devices. The new way seems more predictable and more in line with the usual case of multiple controllers being plugged in (Like if someone comes to join in on a game I'm playing, I don't expect to suddenly be player 2). Granted I don't feel particularly strongly about this. If someone wanted to argue this is dumb and have the patch be removed that'd be -okay- with me.

But being able to have some control would really help weird edge cases like those of us who build arcade machines or handhelds or whatever because I can imagine people would might want the opposite behavior on their builds too.

**Comment by dankcushions** [comment]
this issue was raised over on the retropie forums also: https://retropie.org.uk/forum/topic/17276/usb-ports-no-longer-determine-player-number-retroarch-1-7-1 0

so yeah, would be idea to have some kind of option to control this.

**Comment by i30817** [comment]
The only other alternative i can think of for automating both cases (keeping the relative order and get rid of useless inbuilt controllers) is to limit the number of players (which you can do in the input submenu) and anchor the retropad devices to the end of the list.

This is impossible to do automatically though.

**Comment by fr500** [comment]
the only decent solution would be to bind on button press instead of on
enumeration

**Comment by i30817** [comment]
Something like a setting called 'Re-assign controllers' with a time limit that resets on a new controller in the main input menu (not the Binds menus)? 

I can see that working ok, but i'd prefer that the current behavior was keep if the user didn't request a change. 

That is, if the user just kept adding controllers, they're assigned as 'normal' to free retropad slots, but if the user requests a 're-assignment' that scheme saves a new complete order. But if the user adds a *new* controller after it'd be nice if it was just added to the end without fuss.

**Comment by fr500** [comment]
Controllers aren't assigned as 'normal' to free retropad slots.
Every time a new controller is added the whole system is reinited (at least on windows / dinput / xinput) and the pad is added depending on the OS enumeration.

Windows (an linux) enumeration tends to be quite random, so you may have one controller plugged as P1, connect another one and that P1 might become P2 and the new one P1 and so on.

This is just my opinion but you keep suggesting convoluted ideas :P. This doesn't require a setting or an option, just one way that works properly.

The "old behavior" that I don't see how is that it would have changed made no sense either, because on many systems telling which port is 0, 1, 2, 3, 4 etc would be impossible. It was not something that was added as a feature, it may have been a casual result of something in the RetroArch code (or even in the linux kernel they are using or whatever), since it was casual and not a predefined codepath it may break / change when anything related changes.

The idea I have for this matter is simple, if you have 10 pads connected, no-pads would be assigned until you press a button. If you press a button in a pad it gets assigned to the first slot, if you press a button in another it would be assigned to the second an so on.

Casually (again) this is the way it works on android, because the OS keeps adding the controllers one after another when they connect.

**Comment by i30817** [comment]
My intent with the suggestion is not force people to configure controllers if the order is predictable to at least some of them right now without that trouble. But i don't even have a controller so i don't personally care about this.

**Comment by i30817** [comment]
Ok, i think i see what you mean now. You want to do that 'user press any button' scheme without a user interface action from the user being necessary.

And any inserted new controller would force all users to decide on a new order.... I can live with that.

**Comment by i30817** [comment]
What happens if all the controllers are assigned and a new one is connected and its button pressed? Does it 'steal' player one?

What i mean is: it's a bit undesirable that on insertion of a new controller every controller needs to get reassigned and can even cause 'stealing' of player number... but if you don't do that, you need a way to reset.

**Comment by fr500** [comment]
> And any inserted new controller would force all users to decide on a new order (?).... I can live with that. I like not having a dedicated menu option.

No, any new inserted controller would be assigned to the next available slot, like what games do, in mario kart wiiu noone is p1. Then one of the users presses a button and he's P1, next one that presses a button is P2 and so forth.

For the 1 person who wants to use more than 16 controllers at the same time I don't have a solution.

**Comment by i30817** [comment]
The options as far as i can tell are (if a save to cfg and a new Input menu option are unneeded):

1. never reassign after reassignment (obviously impossible)
2. reassign all controllers every time when a new controller is inserted transparently by keypress.
3. reassign _once_ on startup by keypress. Require closing down RA in order to change ordering (for instance to use wiimote instead of a normal controller as player 1, if RA ever gets wiimote support).

So you're saying option 3?

**Comment by fr500** [comment]
Yeah but I don{t see why it would require closing down RA, every subsystem is reinited in content load so whenever you load a game all ports are free again.

**Comment by i30817** [comment]
Oh. Ok, that's fine. Only a problem if they try to change when the games are running and you can just tell 'not supported' (or figure out a fix).

**Comment by i30817** [comment]
Does this work for keyboards too?

**Comment by fr500** [comment]
no, keyboard keys are hardcoded to the port they are assigned to and I don't mean to change that at all.
Actually this is just my idea for the functionality but I have no idea about how to accomplish this.

As I said it works like this on android, but casually, not because it was designed like this.

**Comment by i30817** [comment]
I'm fine with the keyboard mappings (if they exist) to be a 'always on' alias for player 1. That's how i (and anyone without a controller) use them anyway.

**Comment by Kayinnasaki** [comment]
As a user, assign on button press seems pretty solid. I can see it confusing some people at first but the functionality seems real good and can accommodate... well, pretty much every weird usecase. Priority is nice for making fixed setups but this is a solution that requires absolutely no cfg tinkering, so that's a huge plus. Very glad to see this issue get discussed seriously, especially since I was unsure if I should even make a feature request about it.

**Comment by i30817** [comment]
It'll probably also fix support for usecase of dolphin Wii games that support Classical gamepad or wiimotes.

Or maybe even psx games with psx mouse support? That'd be cool, though i'm thinking it might not be possible to recognize them because the core wouldn't have anyway to know the game accepts mouse.

**Comment by chewvodka** [comment]
I'm not sure how this would work in my case.  I use a wireless Xbox 360 controller as my primary Player 1 controller.  I have the option, on occasion, to use legit NES controllers for my NES games.  They're currently connected through an adapter to port #1 on my USB.  I also have the option, on occasion, to use legit 6-button Sega controllers for Sega games, and/or legit Atari joysticks for Atari systems.  They connect through an adapter to port #2 on my USB.  Port #3 is a hard drive.  Port #4 is actually going to a powered USB HUB, where I have the Xbox dongle connected to "#5" and a wifi Keyboard/Mouse dongle is in "#6".

The way my system works currently, the Xbox controller is the only one I have mapped in RA, and functions as my constant Player 1.  Through the use of a special driver (XBOXDRV) I essentially map the other controllers to keyboard keys as needed during RA loading.

When the new release of RA came out with this Controller Priority issue, I lost all functionality of my Xbox controller inside RA.  This is because it is plugged into port "#5".  None of my other controllers work either due to them not actually being mapped inside RA.  The first thing I tried to do though was to use a keyboard, since XBOXDRV was mapping those other controllers to keyboard.  It didn't work either.  I suspect now that it was because the keyboard is connected to port "#6".

In essence.. this entire patch edit destroyed my setup.  It now requires me to take my whole system apart, reconnect my keyboard dongle to port #1, the xbox to port #2, and then re-program all of my "on-launch" directions for the XBOXDRV.  I have it the way it is now, because it was the only way to make it work 100% of the time whether my Xbox controller was on or off during power on.  Why not just leave it the way it was?  I mean, was there some urgent and immediate need to add this to RA?

PS, I like the idea of a "press a key to assign player 1" pop up on screen, but I don't see how it can be done to work with all scenarios.  Like in my case, if I want to use a Sega controller one time, an NES another, and Xbox all other times.

**Comment by fr500** [comment]
```
every subsystem is reinited in content load so whenever you load a game all ports are free again.
```

You would just have to press the button of the pad you want to use

**Comment by twinaphex** [comment]
> PS, I like the idea of a "press a key to assign player 1" pop up on screen, but I don't see how it can be done to work with all scenarios. Like in my case, if I want to use a Sega controller one time, an NES another, and Xbox all other times.

Well, it is already done like this on Android. 

I like the idea of having a setting that would allow you to make the input autodetection work like this for other platforms as well.

**Comment by orbea** [comment]
Another possible idea would be able to blacklist certain input devices so RetroArch would ignore them.

**Comment by clebercasali** [comment]
This is related: https://github.com/libsdl-org/SDL/issues/4430

**Comment by clebercasali** [comment]
SDL fixed a similar issue with this commit: https://github.com/libsdl-org/SDL/commit/ba8bc143c1691a837a9265d1911ea737d7bb8002

**Comment by clebercasali** [comment]
This is broken again in SDL.
"env SDL_JOYSTICK_DISABLE_UDEV=1 retroarch" fixes it if using SDL2 driver for controllers.
